README.md
# GoPrecure
goprecure is golang implementation of [Go! Princess PreCure](https://en.wikipedia.org/wiki/Go!_Princess_PreCure)

[![Build Status](https://travis-ci.org/sue445/goprecure.svg?branch=master)](https://travis-ci.org/sue445/goprecure)
[![Coverage Status](https://coveralls.io/repos/sue445/goprecure/badge.svg?branch=master&service=github)](https://coveralls.io/github/sue445/goprecure?branch=master)

## Install
```sh
$ go get github.com/sue445/goprecure
```

## Usage
via. [gore](https://github.com/motemen/gore)

```go
gore> :import github.com/sue445/goprecure
gore> flora := goprecure.NewCureFlora()
gore> flora.Name()
"春野はるか"

gore> flora.Transform()
プリキュア！プリンセスエンゲージ！
咲き誇る花のプリンセス！キュアフローラ！
強く、やさしく、美しく！
Go!プリンセスプリキュア！
冷たい檻に閉ざされた夢、返していただきますわ！
お覚悟はよろしくて？

gore> flora.Name()
"キュアフローラ"

gore> flora.Exchange(goprecure.Flora)
エクスチェンジ！モードエレガント！
舞え、花よ！プリキュア・フローラル・トルビヨン！
ごきげんよう

gore> ret, err = flora.Exchange(goprecure.Mermaid)
gore> err
&errors.errorString{
  s: "キュアフローラ can not use Mermaid",
}
```

## Demo
### Transform
![flora_transform.gif](img/flora_transform.gif)

[demo/flora_transform.go](demo/flora_transform.go)

### Exchange
![flora_exchange.gif](img/flora_exchange.gif)

[demo/flora_exchange.go](demo/flora_exchange.go)

## More examples
* [cure_flora_test.go](cure_flora_test.go)
* [cure_mermaid_test.go](cure_mermaid_test.go)
* [cure_twinkle_test.go](cure_twinkle_test.go)
* [cure_scarlet_test.go](cure_scarlet_test.go)

## Development

```sh
$ go get github.com/mattn/gom
$ gom install

$ export GOPATH="`pwd`/_vendor":$GOPATH
$ _vendor/bin/gore
```

## License
MIT


CHANGELOG.md
# Change Log

## [0.1.1](https://github.com/sue445/goprecure/tree/0.1.1) (2015-12-07)
[Full Changelog](https://github.com/sue445/goprecure/compare/0.1.0...0.1.1)

**Merged pull requests:**

- Add go vet [\#4](https://github.com/sue445/goprecure/pull/4) ([sue445](https://github.com/sue445))
- fix typo [\#3](https://github.com/sue445/goprecure/pull/3) ([sue445](https://github.com/sue445))

## [0.1.0](https://github.com/sue445/goprecure/tree/0.1.0) (2015-12-06)
**Merged pull requests:**

- Setup coveralls [\#2](https://github.com/sue445/goprecure/pull/2) ([sue445](https://github.com/sue445))
- Setup travis [\#1](https://github.com/sue445/goprecure/pull/1) ([sue445](https://github.com/sue445))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

girl_test.go
package goprecure

import (
	"fmt"
	"testing"
)

func ExampleGirl_Name() {
	flora := NewCureFlora()
	flora.PrintLine = false

	// human
	fmt.Println("Name", flora.Name())

	flora.Transform()

	// precure
	fmt.Println("Name", flora.Name())

	// Output:
	// Name 春野はるか
	// Name キュアフローラ
}

func TestTransform(t *testing.T) {
	flora := NewCureFlora()
	flora.PrintInterval = 0
	flora.PrintLine = false

	if flora.Name() != "春野はるか" {
		t.Errorf("actual=%s, expect=%s", flora.Name(), "春野はるか")
	}

	flora.Transform()

	if flora.Name() != "キュアフローラ" {
		t.Errorf("actual=%s, expect=%s", flora.Name(), "キュアフローラ")
	}

	flora.Transform()

	if flora.Name() != "春野はるか" {
		t.Errorf("actual=%s, expect=%s", flora.Name(), "春野はるか")
	}
}

func ExampleGirl_Transform() {
	flora := NewCureFlora()
	flora.PrintInterval = 0

	// human
	fmt.Println("Name", flora.Name())

	flora.Transform()

	// precure
	fmt.Println("Name", flora.Name())

	// Output:
	// Name 春野はるか
	// プリキュア！プリンセスエンゲージ！
	// 咲き誇る花のプリンセス！キュアフローラ！
	// 強く、やさしく、美しく！
	// Go!プリンセスプリキュア！
	// 冷たい檻に閉ざされた夢、返していただきますわ！
	// お覚悟はよろしくて？
	// Name キュアフローラ
}

func TestHumanize(t *testing.T) {
	flora := NewCureFlora()
	flora.PrintInterval = 0
	flora.PrintLine = false

	flora.Humanize()

	if flora.Name() != "春野はるか" {
		t.Errorf("actual=%s, expect=%s", flora.Name(), "春野はるか")
	}

	flora.Transform()

	flora.Humanize()

	if flora.Name() != "春野はるか" {
		t.Errorf("actual=%s, expect=%s", flora.Name(), "春野はるか")
	}
}

func ExampleGirl_Humanize() {
	flora := NewCureFlora()
	flora.PrintLine = false

	// human
	fmt.Println("Name", flora.Name())

	flora.Transform()

	// precure
	fmt.Println("Name", flora.Name())

	flora.Humanize()

	// human
	fmt.Println("Name", flora.Name())

	// Output:
	// Name 春野はるか
	// Name キュアフローラ
	// Name 春野はるか
}

func TestExchange(t *testing.T) {
	flora := NewCureFlora()
	flora.PrintLine = false

	flora.Transform()

	flora.Exchange(Flora)
}

func ExampleGirl_Exchange_human() {
	flora := NewCureFlora()

	_, err := flora.Exchange(Flora)
	fmt.Println(err)

	// Output:
	// Human can not exchange!
}

func ExampleGirl_Exchange_precure() {
	flora := NewCureFlora()
	flora.PrintInterval = 0
	flora.PrintLine = false

	flora.Transform()

	flora.PrintLine = true
	flora.Exchange(Flora)

	// Output:
	// エクスチェンジ！モードエレガント！
	// 舞え、花よ！プリキュア・フローラル・トルビヨン！
	// ごきげんよう
}

func ExampleGirl_Exchange_invalidDressupKey() {
	flora := NewCureFlora()
	flora.PrintInterval = 0
	flora.PrintLine = false

	flora.Transform()

	_, err := flora.Exchange(Mermaid)
	fmt.Println(err)

	// Output:
	// キュアフローラ can not use Mermaid
}

func TestCanUseKey(t *testing.T) {
	flora := NewCureFlora()

	actual := flora.CanUseKey(Flora)

	if !actual {
		t.Errorf("expect flora.canUseKey(Flora) == true")
	}

	actual = flora.CanUseKey(Mermaid)

	if actual {
		t.Errorf("expect flora.canUseKey(Mermaid) == false")
	}
}

func ExampleGirl_CanUseKey() {
	flora := NewCureFlora()

	ret1 := flora.CanUseKey(Flora)
	fmt.Println("flora can use Flora ?:", ret1)

	ret2 := flora.CanUseKey(Mermaid)
	fmt.Println("flora can use Mermaid ?:", ret2)

	// Output:
	// flora can use Flora ?: true
	// flora can use Mermaid ?: false
}

func ExampleNewCureFlora() {
	flora := NewCureFlora()

	fmt.Println("HumanName", flora.HumanName)
	fmt.Println("PrecureName", flora.PrecureName)
	fmt.Println("CastName", flora.CastName)
	fmt.Println("Color", flora.Color)
	fmt.Println("CreatedDate", flora.CreatedDate)
	fmt.Println("DressupKeys", flora.DressupKeys)

	// Output:
	// HumanName 春野はるか
	// PrecureName キュアフローラ
	// CastName 嶋村侑
	// Color pink
	// CreatedDate 2015-02-01
	// DressupKeys [Flora Rose Lily Sakura Royal]
}

func ExampleNewCureMermaid() {
	mermaid := NewCureMermaid()

	fmt.Println("HumanName", mermaid.HumanName)
	fmt.Println("PrecureName", mermaid.PrecureName)
	fmt.Println("CastName", mermaid.CastName)
	fmt.Println("Color", mermaid.Color)
	fmt.Println("CreatedDate", mermaid.CreatedDate)
	fmt.Println("DressupKeys", mermaid.DressupKeys)

	// Output:
	// HumanName 海藤みなみ
	// PrecureName キュアマーメイド
	// CastName 浅野真澄
	// Color blue
	// CreatedDate 2015-02-08
	// DressupKeys [Mermaid Ice Bubble Sango Royal]
}

func ExampleNewCureTwinkle() {
	twinkle := NewCureTwinkle()

	fmt.Println("HumanName", twinkle.HumanName)
	fmt.Println("PrecureName", twinkle.PrecureName)
	fmt.Println("CastName", twinkle.CastName)
	fmt.Println("Color", twinkle.Color)
	fmt.Println("CreatedDate", twinkle.CreatedDate)
	fmt.Println("DressupKeys", twinkle.DressupKeys)

	// Output:
	// HumanName 天ノ川きらら
	// PrecureName キュアトゥインクル
	// CastName 山村響
	// Color yellow
	// CreatedDate 2015-02-22
	// DressupKeys [Twinkle Luna ShootingStar Ginga Royal]
}

func ExampleNewCureScarlet() {
	scarlett := NewCureScarlet()

	fmt.Println("HumanName", scarlett.HumanName)
	fmt.Println("PrecureName", scarlett.PrecureName)
	fmt.Println("CastName", scarlett.CastName)
	fmt.Println("Color", scarlett.Color)
	fmt.Println("CreatedDate", scarlett.CreatedDate)
	fmt.Println("DressupKeys", scarlett.DressupKeys)

	// Output:
	// HumanName 紅城トワ
	// PrecureName キュアスカーレット
	// CastName 沢城みゆき
	// Color red
	// CreatedDate 2015-07-05
	// DressupKeys [Scarlet Hanabi Phoenix Sun Royal]
}


cure_flora_test.go
package goprecure

import (
	"fmt"
)

func ExampleCureFlora() {
	flora := NewCureFlora()
	flora.PrintInterval = 0

	// human
	fmt.Println("Name", flora.Name())

	flora.Transform()

	// precure
	fmt.Println("Name", flora.Name())

	fmt.Println(Flora.String())
	flora.Exchange(Flora)

	fmt.Println(Rose.String())
	flora.Exchange(Rose)

	fmt.Println(Lily.String())
	flora.Exchange(Lily)

	fmt.Println(Sakura.String())
	flora.Exchange(Sakura)

	// Output:
	// Name 春野はるか
	// プリキュア！プリンセスエンゲージ！
	// 咲き誇る花のプリンセス！キュアフローラ！
	// 強く、やさしく、美しく！
	// Go!プリンセスプリキュア！
	// 冷たい檻に閉ざされた夢、返していただきますわ！
	// お覚悟はよろしくて？
	// Name キュアフローラ
	// Flora
	// エクスチェンジ！モードエレガント！
	// 舞え、花よ！プリキュア・フローラル・トルビヨン！
	// ごきげんよう
	// Rose
	// ローズ！
	// 舞え、薔薇よ！プリキュア・ローズ・トルビヨン！
	// ごきげんよう
	// Lily
	// リリィ！
	// 舞え、百合よ！プリキュア・リース・トルビヨン！
	// ごきげんよう
	// Sakura
	// モードエレガント！
	// 桜！珊瑚！銀河！サン！
	// ドレスアッププレミアム！
	// はぁッ！
	// 響け全ての力！プリキュアエクラエスポワール！
	// ごきげんよう
}


dressup_key_test.go
package go_princess_precure

import (
	"testing"
)

func TestCanUseKey(t *testing.T) {
	flora := NewCureFlora()

	actual := flora.CanUseKey(TransformFlora)

	if !actual {
		t.Errorf("expect flora.canUseKey(TransformFlora) == true")
	}

	actual = flora.CanUseKey(TransformMermaid)

	if actual {
		t.Errorf("expect flora.canUseKey(TransformMermaid) == false")
	}
}


girl.go
package goprecure

import (
	"errors"
	"fmt"
	"strings"
	"time"
)

const (
	DefaultIntervalSec = 1
)

type Girl struct {
	HumanName        string
	PrecureName      string
	CastName         string
	Color            string
	CreatedDate      string
	TransformMessage string
	DressupKeys      []DressupKey
	PrintInterval    time.Duration
	PrintLine        bool
	currentState     int
}

func (g *Girl) Name() string {
	switch g.currentState {
	case 0:
		return g.HumanName
	case 1:
		return g.PrecureName
	}
	return g.HumanName
}

func (g *Girl) Transform() *Girl {
	if g.currentState == 0 {
		g.printByLine(g.TransformMessage)
	}

	g.currentState++

	if g.currentState > 1 {
		g.currentState = 0
	}

	return g
}

func (g *Girl) printByLine(message string) {
	if !g.PrintLine {
		return
	}

	for _, line := range strings.Split(message, "\n") {
		line = strings.TrimSpace(line)
		fmt.Println(line)
		time.Sleep(g.PrintInterval)
	}
}

func (g *Girl) Humanize() *Girl {
	g.currentState = 0
	return g
}

func (g *Girl) Exchange(k DressupKey) (*Girl, error) {
	if g.currentState == 0 {
		return g, errors.New("Human can not exchange!")
	}

	if !g.CanUseKey(k) {
		return g, errors.New(g.PrecureName + " can not use " + k.String())
	}

	var message = exchangeMessage(k)
	g.printByLine(message)

	return g, nil
}

func (g *Girl) CanUseKey(k DressupKey) bool {
	for _, value := range g.DressupKeys {
		if value == k {
			return true
		}
	}
	return false
}

func newGirl() *Girl {
	g := new(Girl)
	g.currentState = 0
	g.PrintInterval = DefaultIntervalSec * time.Second
	g.PrintLine = true
	return g
}

func NewCureFlora() *Girl {
	g := newGirl()
	g.HumanName = "春野はるか"
	g.PrecureName = "キュアフローラ"
	g.CastName = "嶋村侑"
	g.Color = "pink"
	g.CreatedDate = "2015-02-01"
	g.DressupKeys = []DressupKey{Flora, Rose, Lily, Sakura, Royal}

	g.TransformMessage = `プリキュア！プリンセスエンゲージ！
咲き誇る花のプリンセス！キュアフローラ！
強く、やさしく、美しく！
Go!プリンセスプリキュア！
冷たい檻に閉ざされた夢、返していただきますわ！
お覚悟はよろしくて？`

	return g
}

func NewCureMermaid() *Girl {
	g := newGirl()
	g.HumanName = "海藤みなみ"
	g.PrecureName = "キュアマーメイド"
	g.CastName = "浅野真澄"
	g.Color = "blue"
	g.CreatedDate = "2015-02-08"
	g.DressupKeys = []DressupKey{Mermaid, Ice, Bubble, Sango, Royal}

	g.TransformMessage = `プリキュア！プリンセスエンゲージ！
澄み渡る海のプリンセス！キュアマーメイド！
強く、やさしく、美しく！
Go!プリンセスプリキュア！
冷たい檻に閉ざされた夢、返していただきますわ！
お覚悟はよろしくて？`

	return g
}

func NewCureTwinkle() *Girl {
	g := newGirl()
	g.HumanName = "天ノ川きらら"
	g.PrecureName = "キュアトゥインクル"
	g.CastName = "山村響"
	g.Color = "yellow"
	g.CreatedDate = "2015-02-22"
	g.DressupKeys = []DressupKey{Twinkle, Luna, ShootingStar, Ginga, Royal}

	g.TransformMessage = `プリキュア！プリンセスエンゲージ！
きらめく星のプリンセス！キュアトゥインクル！
強く、やさしく、美しく！
Go!プリンセスプリキュア！
冷たい檻に閉ざされた夢、返していただきますわ！
お覚悟はよろしくて？`

	return g
}

func NewCureScarlet() *Girl {
	g := newGirl()
	g.HumanName = "紅城トワ"
	g.PrecureName = "キュアスカーレット"
	g.CastName = "沢城みゆき"
	g.Color = "red"
	g.CreatedDate = "2015-07-05"
	g.DressupKeys = []DressupKey{Scarlet, Hanabi, Phoenix, Sun, Royal}

	g.TransformMessage = `プリキュア！プリンセスエンゲージ！
深紅の炎のプリンセス！キュアスカーレット！
強く、やさしく、美しく！
Go!プリンセスプリキュア！
冷たい檻に閉ざされた夢、返していただきますわ！
お覚悟決めなさい！`

	return g
}


dressup_key.go
package goprecure

import (
	"fmt"
)

type DressupKey int

const (
	Flora DressupKey = iota
	Mermaid
	Twinkle
	Scarlet
	Rose
	Ice
	Luna
	Hanabi // Fireworks
	Lily
	Bubble
	ShootingStar
	Phoenix
	Sakura // Cherry Blossom
	Sango  // Coral
	Ginga  // Galaxy
	Sun
	Royal
)

func exchangeMessage(k DressupKey) string {
	switch k {
	case Flora:
		return `エクスチェンジ！モードエレガント！
舞え、花よ！プリキュア・フローラル・トルビヨン！
ごきげんよう`

	case Mermaid:
		return `エクスチェンジ！モードエレガント！
高鳴れ、海よ！プリキュア・マーメイド・リップル！
ごきげんよう`

	case Twinkle:
		return `エクスチェンジ！モードエレガント！
キラキラ、星よ！プリキュア・トゥインクル・ハミング！
ごきげんよう`

	case Scarlet:
		return `エクスチェンジ！モードエレガント！
たぎれ、炎よ！プリキュア・スカーレット・フレイム！
ごきげんよう`

	case Rose:
		return `ローズ！
舞え、薔薇よ！プリキュア・ローズ・トルビヨン！
ごきげんよう`

	case Ice:
		return `アイス！
高鳴れ、氷よ！プリキュア・フローズン・リップル！
ごきげんよう`

	case Luna:
		return `ルナ！
キラキラ、月よ！プリキュア・フルムーン・ハミング！
ごきげんよう`

	case Hanabi:
		return `花火！
燃えよ、炎よ！プリキュア・スカーレット・スパーク！
ごきげんよう`

	case Lily:
		return `リリィ！
舞え、百合よ！プリキュア・リース・トルビヨン！
ごきげんよう`

	case Bubble:
		return `バブル！
高まれ、泡よ！プリキュア・バブルリップル！
ごきげんよう`

	case ShootingStar:
		return `シューティングスター！
キラキラ、流れ星よ！プリキュア・ミーティア・ハミング！
ごきげんよう`

	case Phoenix:
		return `スカーレットバイオリン！フェニックス！
羽ばたけ炎の翼！プリキュア・フェニックス・ブレイズ！
ごきげんよう`

	case Sakura:
		fallthrough
	case Sango:
		fallthrough
	case Ginga:
		fallthrough
	case Sun:
		return `モードエレガント！
桜！珊瑚！銀河！サン！
ドレスアッププレミアム！
はぁッ！
響け全ての力！プリキュアエクラエスポワール！
ごきげんよう`

	case Royal:
		return `モードエレガント！ロイヤル！
ドレスアップロイヤル！
はぁっ！
響け遥か彼方へ！プリキュアグランプランタン！
ブルーミング
ごきげんよう`
	}

	panic(fmt.Sprintf("Unknown DressUpKey: %d", k))
}


go_princess_precure.go
package go_princess_precure

import (
	"fmt"
	"strings"
	"time"
)

const (
	DefaultIntervalSec = 1
)

type DressupKey int

const (
	TransformFlora DressupKey = iota
	TransformMermaid
	TransformTwinkle
	TransformScarlett
)

func exchangeMessage(k DressupKey) string {
	switch k {
	case TransformFlora:
		return `エクスチェンジ！モードエレガント！
舞え、花よ！プリキュア・フローラル・トルビヨン！
ごきげんよう`

	case TransformMermaid:
		return `エクスチェンジ！モードエレガント！
高鳴れ、海よ！プリキュア・マーメイド・リップル！
ごきげんよう`

	case TransformTwinkle:
		return `エクスチェンジ！モードエレガント！
キラキラ、星よ！プリキュア・トゥインクル・ハミング！
ごきげんよう`

	}
	panic(fmt.Sprintf("Unknown DressUpKey: ", k.String()))
}

type Girl struct {
	girl_name          string
	human_name         string
	precure_name       string
	cast_name          string
	color              string
	created_date       string
	current_state      int
	transform_interval time.Duration
	transform_messages map[string]string
	dressup_keys       []DressupKey
}

func (g *Girl) Name() string {
	switch g.current_state {
	case 0:
		return g.human_name
	case 1:
		return g.precure_name
	}
	return g.human_name
}

func (g *Girl) Transform() {
	g.printByLine(g.transform_messages["normal"])

	g.current_state++

	if g.current_state > 1 {
		g.current_state = 0
	}
}

func (g *Girl) printByLine(message string) {
	for _, line := range strings.Split(message, "\n") {
		line = strings.TrimSpace(line)
		fmt.Println(line)
		time.Sleep(g.transform_interval)
	}
}

func (g *Girl) Humanize() {
	g.current_state = 0
}

func (g *Girl) Exchange(k DressupKey) {
	if g.current_state == 0 {
		panic("Human can not exchange!")
	}

	if !g.CanUseKey(k) {
		panic(g.precure_name + " can not use " + k.String())
	}

	g.current_state = 2

	var message = exchangeMessage(k)
	g.printByLine(message)
}

func (g *Girl) CanUseKey(k DressupKey) bool {
	for _, value := range g.dressup_keys {
		if value == k {
			return true
		}
	}
	return false
}

func newGirl() *Girl {
	g := new(Girl)
	g.current_state = 0
	g.transform_interval = DefaultIntervalSec * time.Second
	return g
}

func NewCureFlora() *Girl {
	g := newGirl()
	g.girl_name = "cure_flora"
	g.human_name = "春野はるか"
	g.precure_name = "キュアフローラ"
	g.cast_name = "嶋村侑"
	g.color = "pink"
	g.created_date = "2015-02-01"

	g.transform_messages = map[string]string{
		"normal": `プリキュア！プリンセスエンゲージ！
咲き誇る花のプリンセス！キュアフローラ！
強く、やさしく、美しく！
Go!プリンセスプリキュア！
冷たい檻に閉ざされた夢、返していただきますわ！
お覚悟はよろしくて？`,
	}

	g.dressup_keys = []DressupKey{TransformFlora}

	return g
}

func NewCureMermaid() *Girl {
	g := newGirl()
	g.girl_name = "cure_mermaid"
	g.human_name = "海藤みなみ"
	g.precure_name = "キュアマーメイド"
	g.cast_name = "浅野真澄"
	g.color = "blue"
	g.created_date = "2015-02-08"

	g.transform_messages = map[string]string{
		"normal": `プリキュア！プリンセスエンゲージ！
澄み渡る海のプリンセス！キュアマーメイド！
強く、やさしく、美しく！
Go!プリンセスプリキュア！
冷たい檻に閉ざされた夢、返していただきますわ！
お覚悟はよろしくて？`,
	}

	g.dressup_keys = []DressupKey{TransformMermaid}

	return g
}

func NewCureTwinkle() *Girl {
	g := newGirl()
	g.girl_name = "cure_twinkle"
	g.human_name = "天ノ川きらら"
	g.precure_name = "キュアトゥインクル"
	g.cast_name = "山村響"
	g.color = "yellow"
	g.created_date = "2015-02-22"

	g.transform_messages = map[string]string{
		"normal": `プリキュア！プリンセスエンゲージ！
きらめく星のプリンセス！キュアトゥインクル！
強く、やさしく、美しく！
Go!プリンセスプリキュア！
冷たい檻に閉ざされた夢、返していただきますわ！
お覚悟はよろしくて？`,
	}

	g.dressup_keys = []DressupKey{TransformTwinkle}

	return g
}

func NewCureScarlett() *Girl {
	g := newGirl()
	g.girl_name = "cure_scarlett"
	g.human_name = "紅城トワ"
	g.precure_name = "キュアスカーレット"
	g.cast_name = "沢城みゆき"
	g.color = "red"
	g.created_date = "2015-07-05"

	g.transform_messages = map[string]string{
		"normal": `プリキュア！プリンセスエンゲージ！
深紅の炎のプリンセス！キュアスカーレット！
強く、やさしく、美しく！
Go!プリンセスプリキュア！
冷たい檻に閉ざされた夢、返していただきますわ！
お覚悟決めなさい！`,
	}

	g.dressup_keys = []DressupKey{TransformScarlett}

	return g
}
